syntax = "proto3";

package blockstorageproto;

service BlockStorage {
  rpc Ping (PingMessage) returns (PingMessage) {}
  rpc Read (ReadRequest) returns (ReadResponse) {}
  rpc Write (WriteRequest) returns (WriteResponse) {}
  rpc Heartbeat (HeartbeatMessage) returns (HeartbeatMessage) {}
  rpc BackupWrite(BackupWriteRequest) returns (Ack) {}
  rpc TriggerSync(TriggerSyncRequest) returns (Ack){}
  rpc SyncBlock(SyncBlockRequest) returns (Ack){}
  rpc FinishSync(FinishSyncRequest) returns(Ack){}
}

message PingMessage { }

message HeartbeatMessage { }

message ReadRequest {
  uint64 address = 1;
}

message ReadResponse {
  bytes data = 1;
}

message WriteRequest {
  uint64 address = 1;
  bytes data = 2;
}

message WriteResponse { }

message BackupWriteRequest {
  uint64 address = 1;
  bytes data = 2;
}

message TriggerSyncRequest {
  int32 sync_id = 1;
}

message SyncBlockRequest {
  int32 sync_id = 1;
  uint64 address = 2;
  bytes data = 3;
}

message FinishSyncRequest {
  int32 sync_id = 1;
  int32 total_blocks = 2;
}

message Ack { }
